rama-http/src/matcher/subdomain_trie.rs

```diff
@@ -0,0 +1,173 @@
+use trie_rs::TrieBuilder;
+use crate::Request;
+use rama_core::{Context, context::Extensions, matcher::Matcher};
+use rama_net::http::RequestContext;
+use rama_net::address::{Host};
+
+#[derive(Debug, Clone)]
+pub struct SubdomainTrieMatcher {
+    trie: trie_rs::Trie<u8>,
+}
+
+impl SubdomainTrieMatcher {
+    pub fn new<I, S>(domains: I) -> Self
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<str>,
+    {
+        let mut builder = TrieBuilder::new();
+        for domain in domains {
+            let reversed = reverse_domain(domain.as_ref());
+            builder.push(reversed.as_bytes());
+        }
+        let trie = builder.build();
+        Self { trie }
+    }
+
+    pub fn is_match(&self, domain: &str) -> bool {
+        let reversed = reverse_domain(domain);
+        let domain_parts = reversed.split('.');
+
+        let mut prefix = String::new();
+
+        for part in domain_parts {
+            prefix.push_str(part);
+            prefix.push('.');
+
+            if self.trie.exact_match(prefix.as_bytes()) {
+                return true;
+            }
+        }
+        
+        false
+    }
+
+}
+
+fn reverse_domain(domain: &str) -> String {
+    let from = domain.strip_prefix('.').unwrap_or(domain);
+    let mut domain = from.split('.').rev().collect::<Vec<&str>>().join(".");
+    domain.push('.');
+    domain
+}
+
+impl<State, Body> Matcher<State, Request<Body>> for SubdomainTrieMatcher {
+    fn matches(
+        &self,
+        ext: Option<&mut Extensions>,
+        ctx: &Context<State>,
+        req: &Request<Body>,
+    ) -> bool {
+        let host = match ctx.get::<RequestContext>() {
+            Some(req_ctx) => req_ctx.authority.host().clone(),
+            None => {
+                let req_ctx: RequestContext = match (ctx, req).try_into() {
+                    Ok(rc) => rc,
+                    Err(err) => {
+                        tracing::error!(error = %err, "SubdomainTrieMatcher: failed to extract request context");
+                        return false;
+                    }
+                };
+                let host = req_ctx.authority.host().clone();
+                if let Some(ext) = ext {
+                    ext.insert(req_ctx);
+                }
+                host
+            }
+        };
+
+        match host {
+            Host::Name(domain) => {
+                let is_match = self.is_match(domain.as_str());
+                tracing::trace!("SubdomainTrieMatcher: matching '{}' => {}", domain, is_match);
+                is_match
+            }
+            Host::Address(_) => {
+                tracing::trace!("SubdomainTrieMatcher: ignoring numeric address");
+                false
+            }
+        }
+    }
+}
+
+impl FromIterator<String> for SubdomainTrieMatcher {
```

Comment:
You just need to implement this once really, which is the same as your constructor, also no reason to collect:

```rust
impl<S> FromIterator<S> for SubdomainTrieMatcher
    where S: AsRef<str>,
{
   #[inline]
    fn from_iter<I: IntoIterator<Item = S>>(iter: I) -> Self {
        SubdomainTrieMatcher::new(iter)
    }
}
```

With this you can drop the `impl<'a> FromIterator<&'a str> for SubdomainTrieMatcher {` impl block as well